### DISCUSSION TITLE
Further customizability in forms utilizing IncompleteResponseMixin

### DISCUSSION THREAD

**Discussion by codematsing** [jrief/django-formset#163] [Ideas]
I'm not completely sure if this is the proper page to raise this.

Also not sure if this could even be further worked around already in the current impl of the library.

I just figured that from my understangin of how the library works with how it introduces IncompleteResponseMixin for selectize widget functionality, that it can be opened for further dev customizability.

Say I have a usecase where I have:

```python
class RequestItem(model):
    quantity = Integer...
    amount = Integer...
    
    @property
    def total(self):
       return quantity * amount
```

and I want to include a `total` field in my form for purpose that user can view total computation and can be alerted that request item cannot be accepted as the total amount cannot exceed 1000

```python
class RequestItemForm():
    total = IntegerField(disabled=True)
    class Meta:
         model = RequestItem
         fields = ['quantity', 'amount']
         
   def clean_total():
       if self.total > 1000:
           raise ValidationError("cannot exceed 1000")
```

Is there a way that when `quantity` or `amount` is change, it will trigger `RequestItemFormView` to change amount to be calculated on backend?

```python
class RequestItemFormView():
      form_class = RequestItemForm()
      
      def foo(self):
          # callback function
          form = get_form()
          form['total'].value = form['quantity'].value * form['amount'].value
          return form
```

I just feel likes since the form already has capability to send Form.data back to FormView using ajax calls, that we can take advantage of this for more customizable functionality.

something like
```python
class RequestItemForm():
    # if we are following current patterns of the library. I assume this is front-end processing?
    total = IntegerField(disabled=True, form.TextInput(attrs={'df-compute': 'quantity * amount')
    # or add {incomplete:True} to trigger js to send data for backend processing
    quantity= IntegerField(disabled=True, form.TextInput(attrs={incomplete:True})
    amount= IntegerField(disabled=True, form.TextInput(attrs={incomplete:True})
```

**Comment by jrief** [comment]
Yes, especially with the possibility to make partial fetches, as introduced in version 1.5.

Could you please come up with a real-world example which I can add to my testapp. If I have time, I might find a solution for this. Sounds to be an interesting use-case.

**Comment by codematsing** [comment]
# Use Cases

## Order Form

Right now at the top of my head is an order form
There's a lot to work with here since grand_total will need computation of subtotals from sibling formset and add shipping cosy as well

![download (1).png](https://github.com/user-attachments/assets/d44764b9-2308-4419-9dc7-b16d1293c1f1)


```python
Class OrderItemForm
    quantity = ...
    amount = ...
    subtotal = (should derive from quantity*amount)

Class OrderForm:
    shipping = ...
    grand_total = ...


Class OrderFormCollection:
    order_item_formset = ...
    order_item_form = ...
```

## Timesheet Form

Another possibility with added complexity is a timesheet form
There's an added challenge to this if the editable fields are only the start and finish time. Backend will require to prefill the timedelta on hours worked

![image](https://github.com/user-attachments/assets/1fbf0f68-0bff-4e1b-bf4f-c3deedd6fda5)

```python
Class TimesheetForm
    start_time = ...
    end_time = ...
    hours_worked = (should be derived from end_time-start_time)
    hourly_rate = ...
    total = (should be derived from hours_worked and hourly_rate)

Class TimesheetSummary:
    grand_total = ...

Class TimesheetFormCollection:
    time_sheet_form= ...
    time_sheet_summary= ...
```

**Comment by codematsing** [comment]
Out of scope but I've also explored introducing youtube tiptap extension to the richtext widget

Im not very adept with ts / js (I'm basically learning as I study your library. Lol) but I was able to add custom dialog form for embeding youtube by copying some stuff from tiptap and your library.

I'd be happy to add a PR for this for your review and/or improvement if you want to add this in your library.
